components/org.wso2.carbon.identity.notification.push.device.handler/src/main/java/org/wso2/carbon/identity/notification/push/device/handler/impl/DeviceHandlerServiceImpl.java

```diff
@@ -0,0 +1,587 @@
+        if (device.isPresent()) {
+            return device.get();
+        } else {
+            throw handleDeviceHandlerClientException(ERROR_CODE_DEVICE_NOT_FOUND_FOR_USER_ID, userId, null);
+        }
+    }
+
+    @Override
+    public void editDevice(String deviceId, String path, String value) throws PushDeviceHandlerException {
+
+        Device device = getDevice(deviceId);
+        handleEditDevice(device, path, value);
+    }
+
+    @Override
+    public RegistrationDiscoveryData getRegistrationDiscoveryData(String username, String tenantDomain)
+            throws PushDeviceHandlerException {
+
+        RegistrationDiscoveryData registrationDiscoveryData = new RegistrationDiscoveryData();
+
+        registrationDiscoveryData.setUsername(MultitenantUtils.getTenantAwareUsername(username));
+
+        // Generate device ID.
+        String deviceId = UUID.randomUUID().toString();
+        registrationDiscoveryData.setDeviceId(deviceId);
+
+        // Generate registration challenge.
+        String challenge = UUID.randomUUID().toString();
+        registrationDiscoveryData.setChallenge(challenge);
+
+        //Set organization ID and organization name if the user is associated with an organization.
+        try {
+            resolveTenantAndOrganizationInfo(registrationDiscoveryData, tenantDomain);
+        } catch (OrganizationManagementException e) {
+            throw new PushDeviceHandlerServerException("Error occurred while resolving the tenant and organization " +
+                    "information.", e);
+        }
+
+        // Setup push notification feature related endpoints.
+        setupEndpoints(registrationDiscoveryData);
+
+        // Store to cache.
+        DeviceRegistrationContext deviceRegistrationContext = new DeviceRegistrationContext(
+                challenge, username, tenantDomain, false, false);
+        deviceRegistrationContextManager.storeRegistrationContext(deviceId, deviceRegistrationContext, tenantDomain);
+
+        return registrationDiscoveryData;
+    }
+
+    @Override
+    public String getPublicKey(String deviceId) throws PushDeviceHandlerException {
+
+        Optional<String> publicKey = deviceDAO.getPublicKey(deviceId);
+        if (publicKey.isPresent()) {
+            return publicKey.get();
+        } else {
+            throw handleDeviceHandlerClientException(ERROR_CODE_PUBLIC_KEY_NOT_FOUND, deviceId, null);
+        }
+    }
+
+    /**
+     * Resolve the tenant and organization information.
+     *
+     * @param registrationDiscoveryData Registration discovery data.
+     * @param tenantDomain              Tenant domain.
+     * @throws OrganizationManagementException Organization Management Exception.
+     */
+    private void resolveTenantAndOrganizationInfo(RegistrationDiscoveryData registrationDiscoveryData,
+                                                  String tenantDomain) throws OrganizationManagementException {
```

Comment:
Rename `tenantDomain` to something like identifier as this can contain either the tenantDomain or the Org uuid (in sub org cases). Update the param description in the comment mentioning this.